internal/exec/engine.go

```diff
@@ -138,6 +140,7 @@ func (e Engine) Run(stageName string) error {
 		}
 		return err
 	}
+	stage.Close()
```

Comment:
> That said, I've been thinking that Ignition should leave the LUKS devices open. This is equivalent to e.g. leaving the RAIDs we created activated. But more importantly, this will be important for root-on-LUKS, where we just want to rely on Ignition on first boot to set up the root block device and the OS doesn't have to worry about figuring out that it needs to open it back up if it's LUKS.

Leaving the devices open makes it so that there would be differences between the first boot & subsequent boots though. The thought I was going with for root-on-LUKS is that a unit would unlock only the root device (ideally the same unit we use on subsequent boots) after ignition disks is ran.

Reply:
For RHCOS we made the decision to close and then re-open it for the simple reason that it checks that we can open the device. If the binding to Clevis fails (or there is a bug) then data could be embargoed behind an unknown password. I'd prefer we at least have _some_ testing mechanism.

Reply:
Ahh, OK. Yeah, I agree with the principle of "make first boot and subsequent boots identical" when it comes to the real root, but not when it comes to the initrd stage. By nature, they are pretty different today and are going to keep being different. E.g. for subsequent boots, we'll be relying on kargs to find the rootfs (see https://github.com/coreos/fedora-coreos-config/pull/503) which of course we can't rely on on firstboot since the systemd generator that picks it up already ran.

I guess we could special-case the rootfs, and close all the other ones since indeed real root should be able to open them the same way they'd be opened on subsequent boot. But then it means that Ignition knows about the root filesystem, which so far it mostly doesn't. So not sure... still mostly leaning towards leaving the devices open?

I think the API I want to establish is that after `ignition-disks` ran, there is a block device labeled `root` which `ignition-ostree-mount-firstboot-sysroot.service` can just mount. That way, the OS doesn't really have to learn how to read Ignition configs (we already sadly do too much of that). And then the rootmap code figures out what kargs are needed to find the rootfs we just mounted on subsequent boots.

Reply:
> For RHCOS we made the decision to close and then re-open it for the simple reason that it checks that we can open the device. If the binding to Clevis fails (or there is a bug) then data could be embargoed behind an unknown password. I'd prefer we at least have some testing mechanism.

Bahh sorry, your comment didn't show up until I submitted mine.

I think if we can find a good solution that doesn't involve not-Ignition trying to parse the Ignition config, it could work. You had a suggestion in https://github.com/coreos/ignition/pull/960#issuecomment-636095257 to try to run before the generator, but I'm not following... the generator runs early at systemd init time, right?